### DISCUSSION TITLE
Gracefully link HRP and min var

### DISCUSSION THREAD

**Discussion by microprediction** [skfolio/skfolio#3] [Ideas]
Hi,

Responding here to a message on LI from Hugo. 

Thanks for pointing out this package it looks nice. Perhaps I can find time to contribute by grafting in some novel ideas per the links below. 

Peter

https://medium.com/geekculture/schur-complementary-portfolios-fix-hierarchical-risk-parity-28b0efa1f35f
https://github.com/microprediction/precise/blob/main/precise/skaters/portfoliostatic/schurportfactory.py

**Comment by HugoDelatte** [comment]
Hi Peter,
Adding the Schur Complementary Portfolio would be a great idea. I was already familiar with your article and I think it would fit nicely in this library. We could start working on it together. I'll create a dev branch for it and you could start with the backbone of the implementation.

**Comment by HugoDelatte** [comment]
Implementation is done in the feature branch.
* Code is here [here](https://github.com/skfolio/skfolio/blob/6d54eac95912a80811b4601ea2d54820d747b2e7/src/skfolio/optimization/cluster/hierarchical/_schur.py#L30)
* Tests are [here](https://github.com/skfolio/skfolio/blob/6d54eac95912a80811b4601ea2d54820d747b2e7/tests/test_optimization/test_cluster/test_schur.py#L1)

Let's converge on the optimal implementation, then I'll include an example in the doc and release it for the next version.

Some details about the implementation:
* I applied TCO (Tail Call Optimisation) by replacing the recursion  by an iteration and inplace covariance update to reduce the call stack and space complexity.
*  Binary search on gamma is applied to both matrix A and D at the same time and symmetrization is applied at the end of the schur augmentation. This seems to improve the stability of the solution as gamma tends to 1.

It works very nicely as seen below. When `gamma` is 0, it's homogenous to HRP and as `gamma` increases it tends to the Minimum Variance:

![newplot (16)](https://github.com/skfolio/skfolio/assets/50495273/a27e6b3c-ad28-48bc-b671-626c12e50b08)


However it gets a bit tricky when `gamma` gets above 0.7 (red dots):

![newplot (15)](https://github.com/skfolio/skfolio/assets/50495273/4ae21bb6-b1be-4e26-b6b1-7376915368a8)

Not sure if this behaviour is expected / how we could improve it?
As I expected only one pareto front as gamma moves from 0 to 1.

Here is the code to produce the graph:

```python
import numpy as np
from skfolio import RiskMeasure, PerfMeasure, Population
from skfolio.datasets import load_sp500_dataset
from skfolio.preprocessing import prices_to_returns
from skfolio.optimization import (
    HierarchicalRiskParity,
    MeanRisk,
    SchurComplementaryAllocation,
)

prices = load_sp500_dataset()
X = prices_to_returns(prices)
X = X["2015":]

hrp = HierarchicalRiskParity()
min_var = MeanRisk()
ptf_hrp = hrp.fit_predict(X)
ptf_min_var = min_var.fit_predict(X)
ptf_hrp.tag = "HRP"
ptf_min_var.tag = "Min Var"

pop = Population([])
for gamma in np.linspace(0,0.7,50):
    schur = SchurComplementaryAllocation(gamma=gamma)
    ptf_schur = schur.fit_predict(X)
    ptf_schur.tag = "Schur 0 to 0.7"
    ptf_schur.name = f"Schur {gamma:0.2f}"
    pop.append(ptf_schur)

for gamma in np.linspace(0.7,1,30):
    schur = SchurComplementaryAllocation(gamma=gamma)
    ptf_schur = schur.fit_predict(X)
    ptf_schur.tag = "Schur 0.7 to 1"
    ptf_schur.name = f"Schur {gamma:0.2f}"
    pop.append(ptf_schur)

pop.append(ptf_hrp)
pop.append(ptf_min_var)

fig = pop.plot_measures(
    x=RiskMeasure.ANNUALIZED_VARIANCE,
    y=PerfMeasure.ANNUALIZED_MEAN
)
fig.show()
```

**Comment by microprediction** [comment]
Hugo,

   I think gamma choices depend very heavily on the problem. TBH it
requires more experiments.

   Thanks for the code this is great. I will aim to add more tests and
examples of using it under known generative models.

Peter

**Comment by HugoDelatte** [comment]
More tests under generative models is a good idea.

I've done some exploration to analyse the stability and convergence of the solution by playing with:
* The **minimum cluster size**: for clusters below this size, we stop using the augmented matrices and fall back to HRP allocation. In the `precise` implementation, the default is 5 (from the function `even_split`)
*  The **propagation coefficient**: we propagate the augmented matrices `A` and `D` to the next allocation step using: $Aaug \times coef + A \times (1-coef)$ and $Daug \times coef + D \times (1-coef)$. The `precise` implementation fully propagate the augmented matrices (coef of 100%).

Below, you can see 16 graphs for different datasets.

We can see that the the convergence tends to decrease when the propagation coefficient increases. For the minimum cluster size, this is less obvious.

To keep it simple, the criteria for convergence is to have a portfolio variance that decreases as gamma increases: $\frac{d \sigma^2(W(\gamma))}{d\gamma}<0$. We could also add a condition on the second derivative.

Unless we are able to find a more elegant way to fix this convergence issue, we could set the **propagation coefficient** with a default value of 0 and implement a convergence test. If that test fail, we would raise an error. In the examples below, that test would start to raise an error for `gamma` above ~0.6-0.9 depending on the dataset. 

Propagation Coef
-------------------
Generated with min cluster size = 0

![newplot(1)](https://github.com/skfolio/skfolio/assets/50495273/f3ee1802-ea6b-47f8-b27e-e49e4e0a82fb)
![newplot(2)](https://github.com/skfolio/skfolio/assets/50495273/c5761ff6-9339-4c9a-baab-52f9aa2a2281)
![newplot(3)](https://github.com/skfolio/skfolio/assets/50495273/fc584ba2-1e93-4f7a-bce9-ca0e3d131ced)
![newplot(4)](https://github.com/skfolio/skfolio/assets/50495273/8aa3c3e8-0dd0-4f8f-94e6-7ea441266245)
![newplot(9)](https://github.com/skfolio/skfolio/assets/50495273/c7ddfa02-5d0f-4bd5-b04e-d0e9b3a45c4b)
![newplot(10)](https://github.com/skfolio/skfolio/assets/50495273/464c802e-46df-4ab3-a238-9bafdef72b6d)
![newplot(11)](https://github.com/skfolio/skfolio/assets/50495273/568cecca-3b2f-42be-9595-b5f53fb3a71a)
![newplot(12)](https://github.com/skfolio/skfolio/assets/50495273/0c6e73a3-9a8a-4b4a-89a8-24593b90184f)

Minimum Cluster Size
------------------------
Generated with propagation coef = 100%

![newplot(5)](https://github.com/skfolio/skfolio/assets/50495273/cdfed9ea-2ee4-42af-9fa6-5f1ce84b2c7a)
![newplot(6)](https://github.com/skfolio/skfolio/assets/50495273/37463b4e-07f0-46c8-b49c-7f7ab1d6c78d)
![newplot(7)](https://github.com/skfolio/skfolio/assets/50495273/f9da9b7c-742e-457a-8e31-43313897a27b)
![newplot(8)](https://github.com/skfolio/skfolio/assets/50495273/023e8c02-7430-405e-89d7-820f480a97c6)
![newplot(13)](https://github.com/skfolio/skfolio/assets/50495273/1297612b-989e-4033-878d-a3ae2840c922)
![newplot(14)](https://github.com/skfolio/skfolio/assets/50495273/75cb24e8-0263-45e3-8f58-68549f532e97)
![newplot(15)](https://github.com/skfolio/skfolio/assets/50495273/2123e5bb-0663-4af8-91ba-1d7f8663e4f8)
![newplot(16)](https://github.com/skfolio/skfolio/assets/50495273/f02e50ee-19cd-4b1c-966a-08efbe27f356)

**Comment by microprediction** [comment]
These plots are really cool

**Comment by microprediction** [comment]
I should play with your code. It should be possible to set up an example
where min var and HRP have the same Sharpe. Then interesting to see the
paths of Schur.

**Comment by microprediction** [comment]
I've resurrected that code and I'm building up from some examples. Working
in branch /issue-3/schur-complementary in my fork.

**Comment by HugoDelatte** [comment]
Now that the [Metadata Routing PR](https://github.com/skfolio/skfolio/pull/64) has been merged, I'll have some time to finalize the Schur Complementary implementation.
@microprediction do you have a preference concerning the solution to fix the convergence issue [mentioned above](https://github.com/skfolio/skfolio/discussions/3#discussioncomment-8818119) or maybe a new idea?

**Reply by microprediction** [reply]
... reading ...

**Comment by HugoDelatte** [comment]
@microprediction  I had another crack at it and found the issue.

As seen in the below two graphs, the condition number of each Schur-augmented block (both 𝐴 and 𝐷) grows with 𝛾 (amount of "off-diagonal" information injected back into the diagonal blocks).

In the examples, we start with the S&P sample covariance matrix which is poorly conditioned (κ≈117). Then around 𝛾≈0.7 one of those block condition numbers blows up and the Pareto frontier lose monotonicity.

In the second graph, I apply shrinkage to the initial covariance first so that 𝜅≈2.  With that well-conditioned starting point, every Schur complement remains PSD and well-conditioned all the way to 𝛾=1, and the Pareto frontier stays monotone.

What happened before is that I was starting with the raw sample covariance (high 𝜅), then during the top-down bisection I’d run a per-block binary search to find the largest local 𝛾 that kept each block PSD. Two bad side-effects arose:
* Local 𝛾 clamping meant each split used a different effective 𝛾, so the overall risk-vs-𝛾 curve fractured into two disjoint Pareto segments.
* Even after clamping, those blocks still had high condition numbers, so although they were barely PSD, they remained numerically unstable.

I also tried regularizing each block in the top-down pass (adding ridges or spectral clipping), but that too broke monotonicity, because I was again injecting different amounts of adjustment at each node.

The solution to preserve a single, monotonic interpolation from HRP (𝛾=0) to full MVO (𝛾=1), is to choose one global adjustment up front. We have two options:
* Pre-regularize the initial covariance so that its condition number is low enough that all downstream Schur steps remain well-conditioned for any 𝛾∈[0,1]. The issue is that we don't know a prior by how much.
* Compute the maximum 𝛾_max that keeps all blocks well-conditioned over the entire hierarchy (a single binary search on the full tree), then clamp the user’s 𝛾≤𝛾_max and never change it per-block.

![newplot (32)](https://github.com/user-attachments/assets/bc293d95-8fbc-4437-af72-047a72eba10b)
![newplot (31)](https://github.com/user-attachments/assets/c301fbf7-6bfe-4448-a8e3-f7e1f18cdc7d)

**Reply by microprediction** [reply]
Super interesting, thanks. Maybe we can chat live about this? 

My first instinct is to try different shrinkage.    "geodesic_interpolation_towards_perfect" for example https://github.com/microprediction/randomcov/blob/main/randomcov/covutil/geodesicinterpolation.py

**Reply by HugoDelatte** [reply]
Yes let's have a live chat.
Different shrinkage --> good idea, I'll try that

**Reply by HugoDelatte** [reply]
I’ve pushed a new version on the feature branch that solves the above.

Here’s what I found after a bit more digging:

* As expected, when the initial covariance is well‐conditioned, we get a smooth, monotonic variance descent from HRP to MVO as γ→1.

* If the initial covariance is ill‐conditioned, the variance curve kinks at some γ turning‐point. The worse the conditioning, the smaller that γ turning‐point.

* Covariance regularization must happen globally (on the initial matrix) to preserve monotonicity. Any local tweak during the top‐down bisection (γ-clamping, shrinkage, tiny ridge, nearest-SPD, etc.) breaks the HRP→MVO monotonic variance descent (think of it as restarting from a new HRP at each split, so you can end up with higher variance at a larger γ than you had at a smaller γ with no adjustment).

* Users already choose how to regularize the initial covariance via our existing estimators. We should emphasize in the docs that global shrinkage (Ledoit–Wolf, factor models, Geodesic --> to be implemented in another PR, etc.) is the right way to get a smooth frontier, and we shouldn’t impose any default here. Let users pick, just like they do for the other allocations methods (MVO, HRP etc.).

* By default, we’ll compute the maximum feasible γ that preserves variance monotonicity (using a bracket+binary‐search algo) and clamp γ to that. We’ll add an option so users can disable or just warn about the cap if they really want to push past the turning point.


WDYT?

**Reply by microprediction** [reply]
This makes a lot of sense. For Schur to converge on MVP it must inherit the same limitations and cautions for ill-conditioned matrices, probably. These bullets are a great usage guide. If, subsequently, somebody is able to try things that relax the need for upfront global conditioning that's cool to but caveat emptor.

**Reply by microprediction** [reply]
And hi to Pascal who is probably the only other person reading this thread :)

**Reply by CarloNicolini** [reply]
Guys, this could very well be a research paper, especially the experiments above are really interesting.
What do you think if we discuss it in more depth?

**Reply by HugoDelatte** [reply]
@CarloNicolini I think that’s a great idea. I’m currently at maximum capacity, so I won’t have much bandwidth for the paper, but I can help with code and numerical simulations.

**Reply by microprediction** [reply]
Actually Rama Cont already agree to publish this: https://arxiv.org/abs/2411.05807 in a math journal. 

I doubt it would be difficult to follow up with a much better empirical paper for a finance journal (and include Hugo's algorithm) and also get that published.

**Comment by microprediction** [comment]
Looking good

**Comment by microprediction** [comment]
@garg-ruchi

**Comment by HugoDelatte** [comment]
Feature implemented in [v0.13.0](https://github.com/skfolio/skfolio/releases/tag/v0.13.0)
